File: apps/mobile/ios/Runner/Native/Camera/CameraHostApiImpl.swift
import AVFoundation
import Flutter
import Foundation
import os.log

private let focusLog = OSLog(subsystem: "com.rarocamera", category: "focus")

final class CameraHostApiImpl: NSObject, CameraHostApi, @unchecked Sendable {
  private let manager = CameraManager()
  private let flutterApi: CameraFlutterApi
  private let thumbnailGenerator = ThumbnailGenerator()
  weak var platformViewFactory: CameraPlatformViewFactory?

  init(messenger: FlutterBinaryMessenger) {
    self.flutterApi = CameraFlutterApi(binaryMessenger: messenger)
    super.init()
    manager.onLensSwitched = { [weak self] lens in
      DispatchQueue.main.async {
        self?.flutterApi.onLensSwitched(lens: lens) { _ in }
      }
    }
    manager.onError = { [weak self] error in
      DispatchQueue.main.async {
        self?.flutterApi.onError(code: error.code, message: error.message) { _ in }
      }
    }
    manager.onFocusResult = { [weak self] point, success in
      DispatchQueue.main.async {
        self?.flutterApi.onFocusChanged(point: point, locked: success) { _ in }
      }
    }
    manager.onRecordingFinished = { [weak self] url, durationMs in
      DispatchQueue.main.async {
        self?.flutterApi.onRecordingFinished(path: url.path, durationMs: Int64(durationMs)) { _ in }
      }
    }
    manager.onRecordingFailed = { [weak self] message in
      DispatchQueue.main.async {
        self?.flutterApi.onRecordingFailed(code: .sessionFailed, message: message) { _ in }
      }
    }
  }

  var cameraManager: CameraManager { manager }

  func discoverCapabilities(completion: @escaping (Result<CameraCapabilities, Error>) -> Void) {
    do {
      let caps = try manager.discoverCapabilities()
      completion(.success(caps))
    } catch let error as CameraNativeError {
      completion(.failure(pigeonError(from: error)))
    } catch {
      completion(.failure(pigeonError(code: .sessionFailed, message: error.localizedDescription)))
    }
  }

  func startSession(
    textureId: Int64,
    config: CameraConfig,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    Task {
      do {
        try await manager.startSession(config: config)
        DispatchQueue.main.async {
          self.flutterApi.onSessionStarted(activeConfig: config) { _ in }
          completion(.success(()))
        }
      } catch let error as CameraNativeError {
        DispatchQueue.main.async {
          completion(.failure(self.pigeonError(from: error)))
        }
      } catch {
        DispatchQueue.main.async {
          completion(.failure(self.pigeonError(code: .sessionFailed, message: error.localizedDescription)))
        }
      }
    }
  }

  func stopSession(completion: @escaping (Result<Void, Error>) -> Void) {
    manager.stopSession()
    flutterApi.onSessionStopped { _ in }
    completion(.success(()))
  }

  func startRecording(options: RecordingOptions) throws -> String {
    let sessionId = UUID().uuidString
    do {
      try manager.startRecording(sessionId: sessionId, codec: options.codec)
      return sessionId
    } catch let error as CameraNativeError {
      throw pigeonError(from: error)
    } catch {
      throw pigeonError(code: .sessionFailed, message: error.localizedDescription)
    }
  }

  func stopRecording() throws {
    do {
      try manager.stopRecording()
    } catch let error as CameraNativeError {
      throw pigeonError(from: error)
    } catch {
      throw pigeonError(code: .sessionFailed, message: error.localizedDescription)
    }
  }

  func generateThumbnail(
    videoPath: String,
    completion: @escaping (Result<String, Error>) -> Void
  ) {
    thumbnailGenerator.generate(videoPath: videoPath) { [weak self] result in
      switch result {
      case let .success(path):
        completion(.success(path))
      case let .failure(error):
        completion(.failure(
          self?.pigeonError(code: .formatUnsupported, message: error.localizedDescription)
            ?? error
        ))
      }
    }
  }

  func switchLens(lens: LensType, completion: @escaping (Result<Void, Error>) -> Void) {
    do {
      try manager.switchLens(lens)
      completion(.success(()))
    } catch let error as CameraNativeError {
      completion(.failure(pigeonError(from: error)))
    } catch {
      completion(.failure(pigeonError(code: .sessionFailed, message: error.localizedDescription)))
    }
  }

  func setFormat(
    resolution: Resolution,
    fps: Fps,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    do {
      try manager.setFormat(resolution: resolution, fps: fps)
      completion(.success(()))
    } catch let error as CameraNativeError {
      completion(.failure(pigeonError(from: error)))
    } catch {
      completion(.failure(pigeonError(code: .formatUnsupported, message: error.localizedDescription)))
    }
  }

  func focusAt(point: FocusPoint, completion: @escaping (Result<Void, Error>) -> Void) {
    let t0 = CFAbsoluteTimeGetCurrent()
    let renderRing: () -> Void = { [weak self] in
      guard let self = self, let factory = self.platformViewFactory else { return }
      factory.lastPlatformView?.showFocusRing(
        at: factory.toViewCoordinates(focusPoint: point)
      )
    }
    let runOnMain: (@escaping () -> Void) -> Void = { block in
      if Thread.isMainThread { block() } else { DispatchQueue.main.async(execute: block) }
    }
    runOnMain {
      renderRing()
      let tRing = CFAbsoluteTimeGetCurrent()
      os_log(
        "focusAt ring rendered ringMs=%.1f input=(%.3f,%.3f)",
        log: focusLog, type: .info,
        (tRing - t0) * 1000, point.x, point.y
      )
      let sensorPoint = self.convertNormalizedToSensor(point) ?? CGPoint(x: point.x, y: point.y)
      let tConvert = CFAbsoluteTimeGetCurrent()
      completion(.success(()))
      self.manager.focusAtAsync(sensorPoint: sensorPoint, normalizedPoint: point) { focusErr in
        let tFocus = CFAbsoluteTimeGetCurrent()
        os_log(
          "focusAt focus applied totalMs=%.1f convertMs=%.1f focusMs=%.1f sensor=(%.3f,%.3f) err=%{public}@",
          log: focusLog, type: .info,
          (tFocus - t0) * 1000, (tConvert - tRing) * 1000, (tFocus - tConvert) * 1000,
          sensorPoint.x, sensorPoint.y,
          focusErr.map { "\($0)" } ?? "nil"
        )
      }
    }
  }

  private func convertNormalizedToSensor(_ point: FocusPoint) -> CGPoint? {
    guard let container = platformViewFactory?.lastPlatformView?.view() as? CameraPreviewContainerView else {
      return nil
    }
    let bounds = container.bounds
    guard bounds.width > 0, bounds.height > 0 else { return nil }
    let layerPoint = CGPoint(
      x: CGFloat(point.x) * bounds.width,
      y: CGFloat(point.y) * bounds.height
    )
    return container.previewLayer.captureDevicePointConverted(fromLayerPoint: layerPoint)
  }

  func requestPermission(completion: @escaping (Result<Bool, Error>) -> Void) {
    Task {
      let granted = await manager.requestPermission()
      DispatchQueue.main.async {
        completion(.success(granted))
      }
    }
  }

  func hasPermission(completion: @escaping (Result<Bool, Error>) -> Void) {
    completion(.success(manager.hasPermission()))
  }

  private func pigeonError(from error: CameraNativeError) -> CameraPigeonError {
    return pigeonError(code: error.code, message: error.message)
  }

  private func pigeonError(code: CameraErrorCode, message: String?) -> CameraPigeonError {
    return CameraPigeonError(code: "\(code)", message: message, details: nil)
  }
}

File: apps/mobile/android/app/src/main/kotlin/com/rarocamera/raro_mobile/camera/CameraHostApiImpl.kt
package com.rarocamera.raro_mobile.camera

import android.os.Handler
import android.os.Looper
import com.rarocamera.raro_mobile.generated.camera.CameraCapabilities
import com.rarocamera.raro_mobile.generated.camera.CameraConfig
import com.rarocamera.raro_mobile.generated.camera.CameraErrorCode
import com.rarocamera.raro_mobile.generated.camera.CameraFlutterApi
import com.rarocamera.raro_mobile.generated.camera.CameraHostApi
import com.rarocamera.raro_mobile.generated.camera.FlutterError
import com.rarocamera.raro_mobile.generated.camera.FocusPoint
import com.rarocamera.raro_mobile.generated.camera.Fps
import com.rarocamera.raro_mobile.generated.camera.LensType
import com.rarocamera.raro_mobile.generated.camera.RecordingOptions
import com.rarocamera.raro_mobile.generated.camera.Resolution

class CameraHostApiImpl(
  private val manager: CameraManager,
  private val flutterApi: CameraFlutterApi,
) : CameraHostApi {
  private val main = Handler(Looper.getMainLooper())

  init {
    manager.onLensSwitched = { lens ->
      main.post { flutterApi.onLensSwitched(lens) {} }
    }
  }

  private fun toFlutterError(e: Throwable): FlutterError {
    if (e is CameraNativeException) {
      return FlutterError(code = e.symbolicCode(), message = e.message, details = null)
    }
    return FlutterError(code = "sessionFailed", message = e.message ?: e.javaClass.simpleName, details = null)
  }

  override fun discoverCapabilities(callback: (Result<CameraCapabilities>) -> Unit) {
    try {
      callback(Result.success(manager.discoverCapabilities()))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun startSession(textureId: Long, config: CameraConfig, callback: (Result<Unit>) -> Unit) {
    try {
      manager.startSession(config)
      main.post { flutterApi.onSessionStarted(config) {} }
      callback(Result.success(Unit))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun stopSession(callback: (Result<Unit>) -> Unit) {
    try {
      manager.stopSession()
      main.post { flutterApi.onSessionStopped {} }
      callback(Result.success(Unit))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun switchLens(lens: LensType, callback: (Result<Unit>) -> Unit) {
    try {
      manager.switchLens(lens)
      callback(Result.success(Unit))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun setFormat(resolution: Resolution, fps: Fps, callback: (Result<Unit>) -> Unit) {
    try {
      manager.setFormat(resolution, fps)
      callback(Result.success(Unit))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun focusAt(point: FocusPoint, callback: (Result<Unit>) -> Unit) {
    try {
      manager.focusAt(point) { locked ->
        main.post { flutterApi.onFocusChanged(point, locked) {} }
      }
      callback(Result.success(Unit))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun startRecording(options: RecordingOptions): String {
    try {
      return manager.startRecording(
        options,
        object : RecordingController.RecordingCallbacks {
          override fun onStarted(sessionId: String) {
            main.post { flutterApi.onRecordingStarted(sessionId) {} }
          }

          override fun onFinished(path: String, durationMs: Long) {
            main.post { flutterApi.onRecordingFinished(path, durationMs) {} }
          }

          override fun onFailed(code: CameraErrorCode, message: String?) {
            main.post { flutterApi.onRecordingFailed(code, message) {} }
          }
        },
      )
    } catch (e: Throwable) {
      throw toFlutterError(e)
    }
  }

  override fun stopRecording() {
    try {
      manager.stopRecording()
    } catch (e: Throwable) {
      throw toFlutterError(e)
    }
  }

  override fun generateThumbnail(videoPath: String, callback: (Result<String>) -> Unit) {
    try {
      callback(Result.success(ThumbnailExtractor.extractFirstFrameJpeg(videoPath)))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun requestPermission(callback: (Result<Boolean>) -> Unit) {
    try {
      callback(Result.success(manager.requestPermission()))
    } catch (e: Throwable) {
      callback(Result.failure(toFlutterError(e)))
    }
  }

  override fun hasPermission(callback: (Result<Boolean>) -> Unit) {
    callback(Result.success(manager.hasPermission()))
  }
}
